From 1bf1984b4fe5f69931ce3d6aae22eea6d51c5c7d Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Fri, 9 Oct 2020 19:45:14 +0800
Subject: [PATCH] DRVLIB_SW-3891:SWITCH_API: Bitmap to indicate usable global
 PCE table

Flow table contains upto 512 programmable rules with two
consecutive ranges,first common region the number of global entries
configurable.This feature provide Bitmap to indicate number of
avaliable free rules in Commom region from TFLOW Table.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 39 ++++++++++++++++++++++
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.h | 15 +++++++--
 .../net/ethernet/lantiq/switch-api/gsw_flow_pce.c  | 20 +++++++++++
 drivers/net/ethernet/lantiq/switch-api/gsw_init.h  |  1 +
 .../ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c |  1 +
 include/net/switch_api/gsw_flow_ops.h              |  4 +++
 include/net/switch_api/lantiq_gsw.h                | 14 ++++++++
 7 files changed, 91 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index 35b3dfa584ec..7f5e79d61532 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -3529,6 +3529,8 @@ void *ethsw_api_core_init(ethsw_core_init_t *ethcinit)
 		/* Number of Entries of Common Region in TFLOW Table */
 		gsw_w32(cdev, PCE_TFCR_NUM_NUM_OFFSET, PCE_TFCR_NUM_NUM_SHIFT,
 			PCE_TFCR_NUM_NUM_SIZE, (reg_val >> 2));
+
+		gsw_global_pcerule_bitmap_alloc(cdev);
 	}
 
 	if (!PrvData->aging_timeout)
@@ -27496,6 +27498,41 @@ GSW_return_t GSW_UnFreeze(void)
 	return 0;
 }
 
+static GSW_return_t GSW_GlobalRule_BitMapFree(void *cdev,
+					       GSW_PCE_GlobalBitMap_t *parm)
+{
+	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
+
+	BUG_ON(!cdev);
+
+	bitmap_clear(gswdev->Global_Rule_BitMap.global_rule_idx_map,
+		   parm->base_index,  parm->num_of_rules);
+
+	return GSW_statusOk;
+}
+
+static GSW_return_t GSW_GlobalRule_BitMapAlloc(void *cdev,
+					       GSW_PCE_GlobalBitMap_t *parm)
+{
+	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
+
+	BUG_ON(!cdev);
+
+	parm->base_index =
+		bitmap_find_next_zero_area(gswdev->Global_Rule_BitMap.global_rule_idx_map,
+					   gswdev->Global_Rule_BitMap.global_rule_map_size,
+					   0, parm->num_of_rules, 0);
+
+	if (unlikely(parm->base_index + parm->num_of_rules >=
+		gswdev->Global_Rule_BitMap.global_rule_map_size))
+		return GSW_statusErr;
+
+	bitmap_set(gswdev->Global_Rule_BitMap.global_rule_idx_map,
+		   parm->base_index,  parm->num_of_rules);
+
+	return GSW_statusOk;
+}
+
 static u8 gsw_search_tflowcontiguousblock
 		(ethsw_api_dev_t *gswdev,
 		u32 blockid, u32 numberofentries)
@@ -27847,6 +27884,8 @@ static void gsw_init_fn_ptrs(struct core_ops *ops)
 	ops->gsw_tflow_ops.TFLOW_PceRuleWrite			= GSW_PceRuleWrite;
 	ops->gsw_tflow_ops.TFLOW_PceRuleAlloc			= GSW_PceRuleAlloc;
 	ops->gsw_tflow_ops.TFLOW_PceRuleFree			= GSW_PceRuleFree;
+	ops->gsw_tflow_ops.TFLOW_GlobalRule_BitMapAlloc = GSW_GlobalRule_BitMapAlloc;
+	ops->gsw_tflow_ops.TFLOW_GlobalRule_BitMapFree  = GSW_GlobalRule_BitMapFree;
 
 	/*QOS operations*/
 	ops->gsw_qos_ops.QoS_MeterCfgGet				= GSW_QoS_MeterCfgGet;
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.h b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.h
index 2619a6a47fc9..845f51ced409 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.h
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.h
@@ -162,9 +162,9 @@ Where as CLEAR_FILL_CTRL_REG takes care of clear then fill.
 /* Pore redirect PCE rules set or port 0 (30), port 1 (31),*/
 /* port 2 (32), port 3(33), port 4(34) and port 5(35) */
 #define PRD_PRULE_INDEX 30
-#define EAPOL_PCE_RULE_INDEX	60
-#define BPDU_PCE_RULE_INDEX 61
-#define MPCE_RULES_INDEX	10
+#define EAPOL_PCE_RULE_INDEX	8
+#define BPDU_PCE_RULE_INDEX 9
+#define MPCE_RULES_INDEX	0
 #define LTQ_GSWIP_2_0 0x100
 #define LTQ_GSWIP_2_1 0x021
 #define LTQ_GSWIP_2_2 0x122
@@ -921,6 +921,14 @@ typedef struct {
 	u32	nIndex;
 } GSW_PCE_RuleManage_t;
 
+/* \brief Bitmap used Global PCE Rules */
+typedef struct {
+	/** Number of Global PCE rules map size. */
+	u32 global_rule_map_size;
+	/** Bitmap for used Global PCE Rules. */
+	unsigned long *global_rule_idx_map;
+} GSW_Global_PCE_RuleMap;
+
 typedef struct {
 	/**Switch Opertations**/
 	struct core_ops ops;
@@ -1040,6 +1048,7 @@ typedef struct {
 	/*IRQ*/
 	struct pce_irq_linklist *PceIrqList;
 	u32  irq_num;
+	GSW_Global_PCE_RuleMap Global_Rule_BitMap;
 } ethsw_api_dev_t;
 
 
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
index 772e7f784d52..87097edb0f67 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
@@ -1754,7 +1754,27 @@ int gsw_pmicro_code_init_f24s(void *cdev)
 	return 0;
 }
 
+int gsw_global_pcerule_bitmap_alloc(void *cdev)
+{
+	ethsw_api_dev_t *gswdev = GSW_PDATA_GET(cdev);
+
+	BUG_ON(!cdev);
+
+	gswdev->Global_Rule_BitMap.global_rule_map_size =
+		 BITS_TO_LONGS(gswdev->num_of_global_rules) *
+		 sizeof(unsigned long) * BITS_PER_BYTE;
+	gswdev->Global_Rule_BitMap.global_rule_idx_map =
+		kzalloc(BITS_TO_LONGS(gswdev->num_of_global_rules) *
+			 sizeof(unsigned long), GFP_KERNEL);
 
+	if (!gswdev->Global_Rule_BitMap.global_rule_idx_map)
+		return -ENOMEM;
+
+	bitmap_set(gswdev->Global_Rule_BitMap.global_rule_idx_map,
+		   MPCE_RULES_INDEX, BPDU_PCE_RULE_INDEX + 1);
+
+	return GSW_statusOk;
+}
 
 /* PCE Table Micro Code Init routine */
 int gsw_pmicro_code_init(void *cdev)
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_init.h b/drivers/net/ethernet/lantiq/switch-api/gsw_init.h
index f2b96d64775c..4d4f7a1aae51 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_init.h
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_init.h
@@ -178,6 +178,7 @@ void gsw_corecleanup(void);
 int	gsw_pmicro_code_init(void *cdev);
 int	gsw_pmicro_code_init_f24s(void *cdev);
 int gsw_swapi_init(void);
+int gsw_global_pcerule_bitmap_alloc(void *cdev);
 
 #if defined(WIN_PC_MODE) && WIN_PC_MODE
 struct core_ops *gsw_get_swcore_ops(u32 devid);
diff --git a/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c b/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c
index 7444158c48c5..ef4c5dfdf6ae 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c
@@ -211,6 +211,7 @@ static int gsw_remove_data(u32 devid)
 					container_of(gsw_dev_cells[i].drv_data,
 						     ethsw_api_dev_t,
 						     ops);
+				kfree(core_pdata->Global_Rule_BitMap.global_rule_idx_map);
 				kfree(core_pdata);
 			}
 		}
diff --git a/include/net/switch_api/gsw_flow_ops.h b/include/net/switch_api/gsw_flow_ops.h
index 562f577b4cf8..1a257ffb36fb 100644
--- a/include/net/switch_api/gsw_flow_ops.h
+++ b/include/net/switch_api/gsw_flow_ops.h
@@ -154,6 +154,10 @@ struct tflow_ops {
 	GSW_return_t (*TFLOW_PceRuleAlloc)(void *, GSW_PCE_rule_alloc_t *);
 	/* Command: GSW_TFLOW_FREE ; Index: 0x05 */
 	GSW_return_t (*TFLOW_PceRuleFree)(void *, GSW_PCE_rule_alloc_t *);
+	/* BitMap to Allocate usuable free Global Entries*/
+	GSW_return_t (*TFLOW_GlobalRule_BitMapAlloc)(void *, GSW_PCE_GlobalBitMap_t *);
+	/* BitMap to Free used Global Entries*/
+	GSW_return_t (*TFLOW_GlobalRule_BitMapFree)(void *, GSW_PCE_GlobalBitMap_t *);
 };
 
 /*QOS operations*/
diff --git a/include/net/switch_api/lantiq_gsw.h b/include/net/switch_api/lantiq_gsw.h
index 92415593a2c5..9aaba63bc150 100644
--- a/include/net/switch_api/lantiq_gsw.h
+++ b/include/net/switch_api/lantiq_gsw.h
@@ -4607,6 +4607,20 @@ typedef struct {
 	u32 blockid;
 } GSW_PCE_rule_alloc_t;
 
+/** \brief TRAFFIC FLOW TABLE  Global BitMap Allocation.
+ */
+typedef struct {
+	/** Number of traffic flow table entries are
+	 * associated to Global or Common Region.
+	 */
+	u32 num_of_rules;
+
+	/** Find a contiguous free Global Rules based on
+	 * num_of_rules assign baseindex
+	 */
+	u32 base_index;
+} GSW_PCE_GlobalBitMap_t;
+
 /*@}*/ /* GSW_IOCTL_GSWIP31 */
 
 /* -------------------------------------------------------------------------- */
